File: 백준/Gold/1238. 파티/파티.py
import sys
input = sys.stdin.readline

from collections import defaultdict, deque
from heapq import heappop, heappush

N, M, X = map(int, input().split(" "))


graph = defaultdict(list)

for _ in range(M):
    u, v, w = map(int, input().split(" "))
    graph[u].append((v, w))




def dijkstra(start):
    pq = []
    dist = [float('inf') for _ in range(N + 1)]

    dist[start] = 0
    heappush(pq, (0, start))

    while pq:
        cur_dist, v = heappop(pq)

        if dist[v] < cur_dist:
            continue

        for next, w in graph[v]:
            new_dist = cur_dist + w

            if dist[next] > new_dist:
                dist[next] = new_dist
                heappush(pq, (new_dist, next))

    return dist


go = dijkstra(X)

for i in range(1, N+1):
    if i == X:
        continue
    come = dijkstra(i)
    go[i] += come[X]

print(max(go[1:]))

File: 백준/Gold/1238. 파티/파티.java
import java.util.*;
import java.io.*;


public class Main {

    static int N, M, X;

    static ArrayList<ArrayList<Node>> graph = new ArrayList<>();

    static int[] dist;


    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        X = Integer.parseInt(st.nextToken());

        for(int i = 0; i<N+1;i++){
            graph.add(new ArrayList<>());
        }



        for(int i = 0; i<M; i++){
            st = new StringTokenizer(br.readLine());

            int u = Integer.parseInt(st.nextToken());
            int v = Integer.parseInt(st.nextToken());
            int w = Integer.parseInt(st.nextToken());

            graph.get(u).add(new Node(v, w));
        }

        int[] answer = dijkstra(X);

        for(int i = 1; i<N+1; i++){
            if(i == X) continue;

            int[] answer2 = dijkstra(i);
            answer[i] += answer2[X];
        }

        int maxVal = 0;
        for(int i = 1; i<N+1; i++){
            if(maxVal < answer[i])
                maxVal = answer[i];
        }

        System.out.println(maxVal);

    }


    static int[] dijkstra(int start){
        dist = new int[N+1];
        Arrays.fill(dist, Integer.MAX_VALUE);

        PriorityQueue<Node> pq = new PriorityQueue<>();

        dist[start] = 0;
        pq.offer(new Node(start, 0));


        while(!pq.isEmpty()){
            Node now = pq.poll();
            int v = now.v;
            int w = now.w;

            if(w > dist[v]){
                continue;
            }


            for(Node next : graph.get(v)){
                int newDist = dist[v] + next.w;

                if(newDist < dist[next.v]){
                    dist[next.v] = newDist;
                    pq.offer(new Node(next.v, newDist));
                }
            }
        }

        return dist;

    }



    static class Node implements Comparable<Node>{
        int v;
        int w;

        public Node(int v, int w){
           this.v = v;
           this.w = w;
        }

        @Override
        public int compareTo(Node other){
            return Integer.compare(this.w, other.w);
        }


    }




}